pkg/mcp/tests/stress/stress_test.go

```diff
@@ -0,0 +1,634 @@
+//  Copyright 2019 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package stress
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"net"
+	"os"
+	"runtime"
+	"sort"
+	"strconv"
+	"sync"
+	"testing"
+	"text/tabwriter"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/golang/protobuf/proto"
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/grpclog"
+
+	mcp "istio.io/api/mcp/v1alpha1"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/mcp/server"
+	"istio.io/istio/pkg/mcp/sink"
+	"istio.io/istio/pkg/mcp/snapshot"
+	"istio.io/istio/pkg/mcp/source"
+	"istio.io/istio/pkg/mcp/testing/groups"
+	"istio.io/istio/pkg/mcp/testing/monitoring"
+)
+
+var (
+	iterations           = flag.Int("iterations", 1000, "number of update iterations")
+	numCollections       = flag.Int("num_collections", 50, "number of collections")
+	numClients           = flag.Int("num_clients", 20, "number of load numClients")
+	randomSeed           = flag.Int64("rand_seed", 0, "initial random seed for generating datasets")
+	numDatasets          = flag.Int("num_datasets", 1000, "number of datasets to pre-generate for the test")
+	maxChangesPerDataset = flag.Int("max_changes", 10, "maximum number of changes per commonDataset")
+	stepRate             = flag.Duration("step_rate", 1*time.Millisecond, "step rate for publishing new snapshot updates")
+	minApplyDelay        = flag.Duration("min_apply_delay", 0, "minimum extra delay inserted during per-client apply.")
+	maxApplyDelay        = flag.Duration("max_apply_delay", 0, "maximum extra delay inserted during per-client apply.")
+	minUpdateDelay       = flag.Duration("min_update_delay", 0, "minimum extra delay inserted between snapshot updates")
+	maxUpdateDelay       = flag.Duration("max_update_delay", 0, "maximum extra delay inserted between snapshot updates")
+	nackRate             = flag.Float64("nack_rate", 0.01, "rate at which clients nack updates. Should be in the range [0,1]")
+	clientIncPercentage  = flag.Float64("client_inc_rate", 0, "percentate of clients that request incremental updates [0,1]")
+)
+
+var (
+	commonDataset []snapshot.Snapshot
+)
+
+func TestMain(m *testing.M) {
```

Comment:
i believe I had an erroneous collection name in the original pack. Would it be too hard to add a stress test for that case as well (i.e. where the client is asking for wrong things)

Reply:
Ah, missed that. Thanks